nvtabular/workflow.py

```diff
@@ -734,83 +718,92 @@ def apply(
-                    )
-            # "Online" apply currently requires manual iteration
+            self.iterate_online(
+                dataset,
+                output_path=output_path,
+                shuffle=shuffle,
+                output_format=output_format,
+                out_files_per_proc=out_files_per_proc,
+            )
+
+    def iterate_online(
+        self,
+        dataset,
+        end_phase=None,
+        output_path=None,
+        shuffle=None,
+        output_format=None,
+        out_files_per_proc=None,
+        apply_ops=True,
+    ):
+        """ Iterate through dataset and (optionally) apply/shuffle/write.
+        """
+        # Check if we have a (supported) writer
+        writer = None
+        writer_cls = None
+        if output_format:
+            if output_format == "parquet":
+                writer_cls = nvt_io.ParquetWriter
+            elif output_format == "hugectr":
+                writer_cls = nvt_io.HugeCTRWriter
+            else:
+                raise ValueError("Output format not yet supported.")
+            output_path = output_path or "./"
+            output_path = str(output_path)
+            fs = get_fs_token_paths(output_path)[0]
+            writer = writer_cls(
+                output_path, num_out_files=out_files_per_proc, shuffle=shuffle, fs=fs
+            )
+
+        # Iterate through dataset, apply ops, and write out processed data
+        if apply_ops:
             for gdf in dataset.to_iter():
-                self.apply_ops(
-                    gdf,
-                    output_path=output_path,
-                    shuffler=shuffler,
-                    num_out_files=out_files_per_proc,
-                    huge_ctr=huge_ctr,
-                )
-            if shuffler:
-                shuffler.close()
-            if huge_ctr:
-                huge_ctr.close()
+                self.apply_ops(gdf, output_path=output_path, writer=writer, shuffle=shuffle)
 
-    def reorder_tasks(self, end):
-        if end != 2:
-            # Opt only works for two phases (for now)
-            return
-        stat_tasks = []
-        trans_tasks = []
-        for idx, _ in enumerate(self.phases[:end]):
-            for task in self.phases[idx]:
-                deps = task[2]
-                if isinstance(task[0], StatOperator):
-                    if deps == ["base"]:
-                        stat_tasks.append(task)
-                    else:
-                        # This statistics depends on a transform
-                        # (Opt wont work)
-                        return
-                elif isinstance(task[0], TransformOperator):
-                    trans_tasks.append(task)
+        # Close writer and write general/specialized metadata
+        if writer:
+            general_md, special_md = writer.close()
+            writer_cls.write_special_metadata(special_md, fs, output_path)
+            writer_cls.write_general_metadata(general_md, fs, output_path)
```

Comment:
I know we talked about this a bit offline, but you are right that this needs to be documented a bit better in the code (or just improved in another way)...  We are defining/using these class-level functions to allow the same `writer` class(es) to be used by Dask.   That is, if we want to enable Dask to use the same IO funcitonality, we need to avoid writing shared metadata on every worker, and instead aggregate the metadata on  the client process before writing out a single file.